File: scripts/post_pin.py
"""
post_pin.py
───────────
Runs daily (10am UTC) via GitHub Actions.

1. Gets next approved, unposted pin from Supabase
2. Posts it to Pinterest (with auto token refresh on 401)
3. Marks it as posted in Supabase
"""

import os
import requests
from datetime import datetime, timezone
from pinterest_auth import PinterestAuth

# ── Secrets ──────────────────────────────────────────────────
SUPABASE_URL       = os.environ["SUPABASE_URL"]
SUPABASE_KEY       = os.environ["SUPABASE_KEY"]
DEFAULT_BOARD_ID   = os.environ.get("PINTEREST_BOARD_ID", "")

SUPABASE_HEADERS = {
    "apikey":        SUPABASE_KEY,
    "Authorization": f"Bearer {SUPABASE_KEY}",
    "Content-Type":  "application/json",
    "Prefer":        "return=representation"
}

PINTEREST_API = "https://api.pinterest.com/v5"


# ── Supabase helpers ─────────────────────────────────────────

def get_next_pin():
    """Get the oldest approved, unposted pin from Supabase."""
    resp = requests.get(
        f"{SUPABASE_URL}/rest/v1/pins",
        headers=SUPABASE_HEADERS,
        params={
            "approved": "eq.true",
            "posted":   "eq.false",
            "order":    "created_at.asc",
            "limit":    "1"
        }
    )
    resp.raise_for_status()
    pins = resp.json()
    return pins[0] if pins else None


def mark_posted(pin_id, pinterest_pin_id):
    """Mark a pin as posted in Supabase."""
    resp = requests.patch(
        f"{SUPABASE_URL}/rest/v1/pins?id=eq.{pin_id}",
        headers=SUPABASE_HEADERS,
        json={
            "posted":       True,
            "pinterest_id": pinterest_pin_id,
            "posted_at":    datetime.now(timezone.utc).isoformat()
        }
    )
    resp.raise_for_status()
    print(f"Pin {pin_id} marked as posted.")


# ── Pinterest posting ─────────────────────────────────────────

MAX_DESC = 500   # Pinterest API hard limit for description
MAX_ALT  = 500   # Pinterest API hard limit for alt_text


def build_payload(pin):
    """Build the Pinterest pin payload from a Supabase pin row.

    Description is sent as-is (already keyword-optimized by Gemini).
    No hashtags — they're dead on Pinterest and can hurt SEO.
    """
    description = (pin.get("description") or "").strip()[:MAX_DESC]
    link_url    = pin.get("link_url", "")

    # Use per-pin board if set, otherwise fall back to env var
    board_id = pin.get("board_id") or DEFAULT_BOARD_ID
    if not board_id:
        raise ValueError("No board_id on pin and no PINTEREST_BOARD_ID env var set.")

    payload = {
        "board_id":    board_id,
        "title":       pin["title"][:100],
        "description": description,
        "media_source": {
            "source_type": "image_url",
            "url":         pin["image_url"]
        }
    }

    # Alt text for accessibility (Pinterest API field)
    alt_text = (pin.get("alt_text") or "").strip()
    if alt_text:
        payload["alt_text"] = alt_text[:MAX_ALT]

    # Only add link if one exists — Pinterest rejects empty link field
    if link_url:
        payload["link"] = link_url

    return payload


# ── Main ─────────────────────────────────────────────────────

if __name__ == "__main__":
    print("=" * 60)
    print(f"Daily pin post — {datetime.now(timezone.utc).strftime('%Y-%m-%d %H:%M UTC')}")
    print("=" * 60)

    # 1. Get next pin from queue
    pin = get_next_pin()
    if not pin:
        print("No approved pins in queue. Nothing posted today.")
        print("Go to your review app and approve some pins.")
        exit(0)

    print(f"Next pin: id={pin['id']} — {pin['title'][:70]}...")
    print(f"Link URL: {pin.get('link_url') or '(none)'}")

    # 2. Validate image URL
    if not pin.get("image_url"):
        print(f"Pin {pin['id']} has no image URL — skipping.")
        exit(1)

    # 3. Build payload
    payload = build_payload(pin)

    # 4. Post to Pinterest — auto-refresh on 401 handled by PinterestAuth
    print("Posting to Pinterest...")
    auth = PinterestAuth()
    resp = auth.post(f"{PINTEREST_API}/pins", json=payload)

    if not resp.ok:
        print(f"Pinterest error {resp.status_code}: {resp.text}")
        resp.raise_for_status()

    result       = resp.json()
    pinterest_id = result.get("id", "unknown")
    print(f"Posted! Pinterest pin ID: {pinterest_id}")
    print(f"Pin URL: https://pinterest.com/pin/{pinterest_id}/")

    # 5. Mark as posted in Supabase
    mark_posted(pin["id"], pinterest_id)
    print("Done.")

File: scripts/edge_functions/post-pin.ts
import { serve } from "https://deno.land/std@0.168.0/http/server.ts"
import { createClient } from "https://esm.sh/@supabase/supabase-js@2"

const PINTEREST_API = "https://api.pinterest.com/v5"
const MAX_DESC = 500
const MAX_ALT  = 500

serve(async (req) => {
  if (req.method === 'OPTIONS') {
    return new Response('ok', { headers: { 'Access-Control-Allow-Origin': '*', 'Access-Control-Allow-Methods': 'POST', 'Access-Control-Allow-Headers': '*' } })
  }

  console.log("--- Post Pin Triggered ---")

  try {
    const { pin_id } = await req.json()
    console.log("Target Pin ID:", pin_id);
    
    const supabase = createClient(
      Deno.env.get('SUPABASE_URL') ?? '',
      Deno.env.get('SUPABASE_SERVICE_ROLE_KEY') ?? ''
    )

    console.log("Fetching tokens from database...");
    const { data: settings, error: setErr } = await supabase
      .from('settings').select('*').eq('id', 1).single()

    if (setErr || !settings) {
      console.error("Settings fetch error:", setErr);
      throw new Error('Pinterest tokens not found. Connect account in Settings first.');
    }

    const { data: pin, error: pinErr } = await supabase
      .from('pins').select('*').eq('id', pin_id).single()

    if (pinErr || !pin) {
      console.error("Pin fetch error:", pinErr);
      throw new Error('Pin not found');
    }

    const accessToken = settings.pinterest_access_token;
    // Per-pin board → settings fallback → env fallback
    const boardId = pin.board_id || settings.pinterest_board_id || Deno.env.get('PINTEREST_BOARD_ID');

    console.log("Payload Check:");
    console.log("- Token present:", !!accessToken);
    console.log("- Board ID:", boardId, pin.board_id ? "(from pin)" : "(from settings/env)");
    console.log("- Title:", pin.title);

    if (!accessToken || !boardId) {
      throw new Error('Missing Pinterest Access Token or Board ID');
    }

    // Description is sent as-is — already keyword-optimized by Gemini.
    // No hashtags — they're dead on Pinterest and can hurt SEO.
    const description = (pin.description || '').trim().slice(0, MAX_DESC)

    const payload: Record<string, unknown> = {
      board_id: boardId,
      title: pin.title.slice(0, 100),
      description: description,
      media_source: { source_type: "image_url", url: pin.image_url }
    }

    // Alt text for accessibility + SEO signal
    const altText = (pin.alt_text || '').trim()
    if (altText) {
      payload.alt_text = altText.slice(0, MAX_ALT)
    }

    // Only add link if present — Pinterest rejects empty link field
    if (pin.link_url) {
      payload.link = pin.link_url
    }

    console.log("Sending to Pinterest API...");
    console.log("- Description length:", description.length);
    console.log("- Alt text present:", !!altText);

    const resp = await fetch(`${PINTEREST_API}/pins`, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${accessToken}`,
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(payload)
    })

    const result = await resp.json()
    if (!resp.ok) {
      console.error("Pinterest API Error Response:", result);
      throw new Error(result.message || 'Pinterest API rejection');
    }

    console.log("Pin created successfully! Pinterest ID:", result.id);

    await supabase.from('pins').update({
      posted: true, pinterest_id: result.id, posted_at: new Date().toISOString()
    }).eq('id', pin_id)

    console.log("Database updated. Job done.");
    return new Response(JSON.stringify({ success: true, pinterest_id: result.id }), {
      headers: { 'Content-Type': 'application/json', 'Access-Control-Allow-Origin': '*' }
    })

  } catch (err) {
    console.error("Critical Post Error:", err.message);
    return new Response(JSON.stringify({ success: false, error: err.message }), {
      status: 400,
      headers: { 'Content-Type': 'application/json', 'Access-Control-Allow-Origin': '*' }
    })
  }
})
